### PR TITLE
[#16740] Add comprehensive RESP command test coverage

### LINKED ISSUE #15965: Bitfield operations
### Description

[BITCOUNT](https://redis.io/docs/latest/commands/bitcount/)
[BITFIELD](https://redis.io/docs/latest/commands/bitfield/)
[BITFIELD_RO](https://redis.io/docs/latest/commands/bitfield_ro/)
[BITOP](https://redis.io/docs/latest/commands/bitop/)
[BITPOS](https://redis.io/docs/latest/commands/bitpos/)
[GETBIT](https://redis.io/docs/latest/commands/getbit/)
[SETBIT](https://redis.io/docs/latest/commands/setbit/)

### Implementation ideas

_No response_

### LINKED ISSUE #16738: SMOVE returns 0 when moved element already exists in destination
## Description

`SMOVE source destination member` returns `0` instead of `1` when the member being moved already exists in the destination set.

## Root cause

In `SMOVE.java`, the `removeAndAdd` method returns the result of `cache.add(destKey, element)`, which returns `0` when the element is already present in the destination. Per the [Redis SMOVE specification](https://redis.io/commands/smove/), the command should return `1` whenever the element was successfully removed from the source set, regardless of whether it was already present in the destination.

## Steps to reproduce

```
SADD src a b c
SADD dst a x y
SMOVE src dst a
# Returns 0, should return 1
```

## Fix

Change the return in `removeAndAdd` from `return cache.add(destKey, element)` to `return cache.add(destKey, element).thenApply(ignore -> 1L)`.

### LINKED ISSUE #16739: ZRANK/ZREVRANK WITHSCORE returns wrong format
## Description

The `ZRANK` and `ZREVRANK` commands with the `WITHSCORE` option have two issues:

1. The option is matched against `WITHSCORES` (plural) instead of `WITHSCORE` (singular), so `ZRANK key member WITHSCORE` is rejected as a syntax error.
2. When the option does match, the score is serialized as an integer instead of a double, producing incorrect results for non-integer scores.

## Root cause

In `ZRANK.java`:
- `isWithScoresArg()` checks for `WITHSCORES`, but the Redis spec for ZRANK/ZREVRANK uses `WITHSCORE` (no trailing S).
- The response uses `ResponseWriter.ARRAY_INTEGER` which serializes all values as integers, truncating fractional scores.

## Steps to reproduce

```
ZADD myset 1.5 a 2.0 b 3.5 c
ZRANK myset a WITHSCORE
# Should return [0, "1.5"] but either errors or returns [0, 1]
```

## Fix

- Match against `WITHSCORE` using `RespUtil.isAsciiBytesEquals`
- Use a custom serializer that writes rank as integer and score as double

### LINKED ISSUE #16740: RESP server test coverage gaps against Redis upstream test suites
## Description

The RESP server command implementations lack comprehensive test coverage when compared against the official Redis test suites (`tests/unit/type/*.tcl`, `tests/unit/expire.tcl`, `tests/unit/bitops.tcl`). Several edge cases, error conditions, and boundary behaviors are untested.

## Scope

The following command families have coverage gaps identified by comparison with the upstream Redis tests:

- **String commands** (`SET`, `GET`, `MSET`, `MGET`, `APPEND`, `INCR`/`DECR`, `GETRANGE`, `SETRANGE`, `GETEX`, `GETDEL`, `SETNX`, `MSETNX`, `SUBSTR`, `LCS`)
- **List commands** (`LPUSH`, `RPUSH`, `LPOP`, `RPOP`, `LRANGE`, `LTRIM`, `LREM`, `LSET`, `LINDEX`, `LINSERT`, `LMOVE`, `RPOPLPUSH`, `LPOS`, `LMPOP`, `LPUSHX`, `RPUSHX`)
- **Hash commands** (`HSET`, `HGET`, `HDEL`, `HINCRBY`, `HINCRBYFLOAT`, `HMGET`, `HSETNX`, `HGETALL`, `HEXISTS`)
- **Set commands** (`SADD`, `SREM`, `SMEMBERS`, `SINTERCARD`, `SINTER`, `SUNION`, `SDIFF`, `SINTERSTORE`, `SUNIONSTORE`, `SDIFFSTORE`, `SMOVE`, `SPOP`, `SRANDMEMBER`)
- **Sorted set commands** (`ZADD`, `ZINCRBY`, `ZREM`, `ZRANGE`, `ZRANGEBYSCORE`, `ZRANGEBYLEX`, `ZRANGESTORE`, `ZRANK`, `ZREVRANK`, `ZDIFF`, `ZUNIONSTORE`, `ZINTERSTORE`, `ZINTERCARD`, `ZPOPMIN`, `ZPOPMAX`, `ZMPOP`, `ZRANDMEMBER`, `ZMSCORE`, `ZSCAN`)
- **Expire commands** (`EXPIRE`, `PEXPIRE`, `EXPIREAT`, `PEXPIREAT`, `PERSIST`, `TTL`, `PTTL`, `EXPIRETIME`, `PEXPIRETIME`, `SETEX`, `PSETEX`, `GETEX`)
- **Bitfield commands** (`BITFIELD` with `GET`, `SET`, `INCRBY`, `OVERFLOW`)
- **Bit operation commands** (`SETBIT`, `GETBIT`, `BITCOUNT`, `BITPOS`, `BITOP`)

Additionally, `RespSingleNodeTest` contains redundant tests that duplicate coverage already present in the specialized test classes, and `StringCommandsTest` has bugs in its LCS `lcsCasesWithMinLen` data provider and match filter.

### PR DESCRIPTION
**Pull request by tristantarrant**
Tests verified against Redis upstream test suites for each command family. Covers string, list, hash, set, sorted set, bitfield,    bitops, and expire commands. Also removes redundant tests from  RespSingleNodeTest that were duplicated in pecialized test classes  and fixes two pre-existing bugs in StringCommandsTest LCS tests.
Fixes bugs with SMOVE, ZRANK/ZREVRANK WITHSCORE

Closes #15965
Closes #16738
Closes #16739
Closes #16740
Closes #16741

### CODE REVIEW

**Comment by wburns** [comment]
StringCommandsClusteredTest seems to be related.

**Review by wburns — APPROVED** [review]
LGTM once CI is good

**Comment by tristantarrant** [comment]
This is now good

File: server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANK.java [hunk 1]
```diff
@@ -24,6 +25,15 @@
  * @since 15.0
  */
 public class ZRANK extends RespCommand implements Resp3Command {
+   private static final byte[] WITHSCORE = "WITHSCORE".getBytes();
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANK.java]: We should really ensure these are using US_ASCII as the charset. It seems unlikely someone will use something different than UTF-8, but if they use something like UTF-16 all of these will no longer work.

Comment (wburns) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANK.java]: We should probably update all the RESP commands to do this btw.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/string/SET.java [hunk 2]
```diff
@@ -40,7 +40,7 @@ public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
                .stageToReturn(SetOperation.performOperation(handler.cache(), arguments, ts, getName()), ctx, SetResponse.SERIALIZER);
       }
       return handler.stageToReturn(
-            handler.ignorePreviousValuesCache().putAsync(arguments.get(0), arguments.get(1)),
+            handler.cache().putAsync(arguments.get(0), arguments.get(1)),
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/string/SET.java]: This change seems wrong, we don't need the return value for SET command as it only replies with OK.
https://redis.io/docs/latest/commands/set/

Guessing Claude was confused by SETBIT?

**Review by wburns — CHANGES_REQUESTED** [review]

File: server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANK.java (hunk 1 above)
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANK.java]: Done.  Although all the strings use ASCII characters, so it doesn't make any difference.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/string/SET.java (hunk 2 above)
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/string/SET.java]: It found missing delegates in `DecorateedCache` and those confused it with the problem it found in `SETRANGE` type checking.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANK.java (hunk 1 above)
Comment (wburns) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANK.java]: It does, unfortunately. The problem is when the default charset is a non ascii based one, such as UTF-16 then "WITHSCORE" will be 18 bytes long instead of the 9 when ASCII encoded and thus won't match the encoding sent from the client.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/string/SET.java [hunk 3]
```diff
@@ -40,7 +40,7 @@ public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
                .stageToReturn(SetOperation.performOperation(handler.cache(), arguments, ts, getName()), ctx, SetResponse.SERIALIZER);
       }
       return handler.stageToReturn(
-            handler.ignorePreviousValuesCache().putAsync(arguments.get(0), arguments.get(1)),
+            handler.ignorePreviousValuesCache().computeAsync(arguments.get(0), (_, _) -> arguments.get(1)),
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/string/SET.java]: Looks like javadoc generation doesn't like this:

/home/runner/work/infinispan/infinispan/server/resp/src/main/java/org/infinispan/server/resp/commands/string/SET.java:43: error: unnamed variables are not supported in -source 17
Error:              handler.ignorePreviousValuesCache().computeAsync(arguments.get(0), (_, _) -> arguments.get(1)),

**Review by wburns — CHANGES_REQUESTED** [review]

**Comment by jabolina** [comment]
main:

```
Summary:
  throughput summary: 79744.82 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.527     0.080     0.463     0.943     1.743    12.535
```

This:

```
Summary:
  throughput summary: 66533.60 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.618     0.096     0.551     1.055     3.199     7.911
```

I've run the command 2x to warm and the third time to collect the value. Running with 3 nodes:

```
$ redis-benchmark -t set -n 100000 -u valkey://admin:password@localhost:11222/0 --user admin -p password -r 1000000 -3  -h 127.0.0.1 -p 11222 --cluster
```

**Comment by tristantarrant** [comment]
I don't think it's worth protecting against WRONGTYPE in SET for such a performance degradation

**Comment by tristantarrant** [comment]
Highly unlikely that the operator CI cache tests are failing because of this. They weren't failing in previous runs

**Comment by tristantarrant** [comment]
Ok, all green now!

**Comment by wburns** [comment]
Integrated into main, thanks @tristantarrant !
